File: Simple_UDP_Ping/UDPClient.java
package week5;

import java.net.*;

class UDPClient {

    public static void main(String argv[]) throws Exception
    {
	// socket variables
	DatagramSocket clientSocket;
	DatagramPacket sendPacket;
	DatagramPacket receivePacket;
	byte[] receiveData = new byte[1024];
	byte[] sendData = new byte[1024];
	InetAddress IPAddress;

	// client variables
	String serverSentence;
	// command-line arguments
	int port;
	String server;
	
	// CONSTANTS
	String MESSAGE = "PING";
	int TIMEOUT = 1000;
	int PINGS = 10;
	
	// statistic arithmetic variables
	double counter = 0, rtt = 0, RTTa = 0, RTTb = 0, minRTT = 0, maxRTT = 0, totalRTT = 0, averateRTT = 0;

	// process command-line arguments
	if (argv.length < 2) {
	    System.out.println ("Usage: java UDPServer hostname port\n");
	    System.exit (-1);
	}
	server = argv[0];
	port = Integer.parseInt(argv[1]);

	// Create client socket to destination
	clientSocket = new DatagramSocket();
	IPAddress = InetAddress.getByName (server);
	
	// set socket time-out setting
	clientSocket.setSoTimeout(TIMEOUT);
	
	// Get input from user
    //System.out.println("Client ready for input");
	//clientSentence = inFromUser.readLine();
	

	for (int i = 0; i < PINGS; i++) {
		// begin try-catch (PING)
		try {
			String sendingMessage = MESSAGE + " " + i + " " + System.currentTimeMillis() + " CRLF";
			// Get input from user
			sendData = sendingMessage.getBytes();
			// Create packet and send to server
			sendPacket = new DatagramPacket(sendData, sendData.length, IPAddress, port);
			
			// Sent time. from time.time()
			RTTb = System.currentTimeMillis();
			
			clientSocket.send(sendPacket);
			//System.out.println (sendingMessage);

			// Create receiving packet and receive from server
			receivePacket = new DatagramPacket(receiveData, receiveData.length); 
			clientSocket.receive(receivePacket);
			
			// Received time. use time.time()
	        RTTa = System.currentTimeMillis();
			// Round trip time is the difference between sent and received time
	        rtt = RTTa - RTTb;
			
	        
	        //serverSentence = new String(receivePacket.getData(),0, 100);
			serverSentence = new String(receivePacket.getData(), 0, receivePacket.getLength());
			String[] setenceStructures = serverSentence.split(" ");

			System.out.println(setenceStructures[0] + " received from " + receivePacket.getAddress() + 
					": seq#=" + setenceStructures[1] + " time= " + rtt + " ms");
			
			//taking tally of a good PING receipt
			counter++;
			// taking some RTT statistics
			if (i == 0) {
				minRTT = rtt;
			    maxRTT = rtt;
			}
			else { 
	            if (rtt < minRTT)
	                minRTT = rtt;
	            if (rtt > maxRTT)
	                maxRTT = rtt;
			}
	        totalRTT += rtt;   
			
		} catch (Exception e) {
			System.err.println("Request timed out.");
		}
		// end try-catch
	}	
	
	
	//print statistics
	System.out.println();
	System.out.println("--- PING statistics ---");
	System.out.println(PINGS + " packets transmitted, " + counter + " received, " + counter/PINGS*100 + "% packet loss");
	System.out.println("rtt min/avg/max = " + minRTT + " " + totalRTT/counter + " " + maxRTT + " ms" );
	
	
	
	// close the socket
	clientSocket.close();

    } // end main

} // end class

File: Simple_UDP_Ping/UDP_Client.py
import sys, time, re
from socket import *

# Get the server hostname and port as command line arguments
numArg = len(sys.argv)
argv = sys.argv        
host = str(sys.argv[1]) 
port = str(sys.argv[2]) 
timeout = 2 # in second, change as needed
pings = 10

# Create UDP client socket
# Note the use of SOCK_DGRAM for UDP datagram packet
clientsocket = socket(AF_INET, SOCK_DGRAM)
# Set socket timeout as 1 second
clientsocket.settimeout(timeout)
# Command line argument is a string, change the port into integer
port = int(port)  
# Sequence number of the ping message
ptime = 0  
counter = 0
minRTT = 0
maxRTT = 0
totalRTT = 0
# Ping for 10 times
while ptime < pings: 
    ptime += 1
    # Format the message to be sent. 
    # use time.asctime() for currTime
    data = ' '.join((str(x) for x in ("PING", ptime, time.asctime(), "CRLF")))
    try:
        # Sent time. from time.time()
        RTTb = time.time()
        # Send the UDP packet with the ping message
        clientsocket.sendto(data.encode(),(host, port))
        # Receive the server response
        message, address = clientsocket.recvfrom(1024)  
        # Received time. use time.time()
        RTTa = time.time()
        # Display the server response as an output
        
        #split the received message into a dictionary
        splitMessage = message.split()

        ipAddress =  str(address)
        ipAddress = ipAddress.split("'")
        # Round trip time is the difference between sent and received time
        rtt = RTTa - RTTb
        receivedMessage = splitMessage[0]
        sequence = splitMessage[1]        
        print(receivedMessage + " received from " + ipAddress[1] + ": seq#= " + sequence + " time= " + str(rtt*1000) + " ms")
        counter += 1
        if ptime == 1:
            minRTT = rtt
            maxRTT = rtt
        else: 
            if rtt < minRTT:
                minRTT = rtt
            if rtt > maxRTT:
                maxRTT = rtt
        totalRTT += rtt              
    except:
        # Server does not response
        # Assume the packet is lost
        print ("Request timed out.")
    continue

packetLoss = (float(pings - counter) / pings) * 100
averageRTT = float(totalRTT) / counter
print "--- ping statistics ---"
print str(pings) + " packets transmitted, " + str(counter) + " received, " + str(packetLoss) + "% packet loss"
print "rtt  min/avg/max = " + str(minRTT*1000) + " " + str(averageRTT*1000) + " " + str(maxRTT*1000) 
# Close the client socket
clientsocket.close()
 




